### PR TITLE
[create_timepoint] upgraded module to use react

### PR DESCRIPTION
**Pull request by maltheism**
### Brief summary of changes

This pull request removes smarty and upgrades module to use react. This PR removes the smarty templates and adds a react file for create_timepoint. Also it adds one php file for handling the post messages from the user.

Note: I changed how the visit label is handled. It's now done on the frontend and so there is no delay for the user.

### This resolves issue...

- [x] Redmine? #16008

### To test this change...

- [x] Pull the PR and test on your LORIS VM.

### CODE REVIEW

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        /**
+         * User permission verification.
+         */
+        $user       = $request->getAttribute('user');
+        $identifier = $requestBody['identifier'] ?? '';
+        $candidate  =& \Candidate::singleton($identifier);
+        if (!$user->hasPermission('data_entry')
+            && !(in_array(
+                $candidate->getData('RegistrationCenterID'),
+                $user->getData('CenterIDs')
```
Comment (johnsaigle) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]: Please add `true` as a third parameter to `in_array`

File: modules/create_timepoint/php/create_timepoint.class.inc [hunk 2]
```diff
@@ -22,31 +22,24 @@ namespace LORIS\create_timepoint;
...
      *
      * @param \User $user The user whose access is being checked
      *
      * @return bool
+     * @throws \LorisException
      */
     function _hasAccess(\User $user) : bool
     {
-        $candidate =& \Candidate::singleton($this->identifier);
+        $candidate = \Candidate::singleton($this->identifier);
```
Comment (johnsaigle) [inline modules/create_timepoint/php/create_timepoint.class.inc]: Just since you're editing this file anyway can you also add the `true` param to in_array below?

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 3]
```diff
@@ -0,0 +1,283 @@
+<?php
```
[the remark this answers is not in the document]
Comment (maltheism) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: @johnsaigle You mean `<?php declare(strict_types=1);` I believe and thanks i'll add!

**Comment by johnsaigle** [comment]
Needs work on the integration tests

**Comment by maltheism** [comment]
Blocking this PR until #4416 is merged.

File: modules/create_timepoint/php/create_timepoint.class.inc (hunk 2 above)
Comment (maltheism) [reply modules/create_timepoint/php/create_timepoint.class.inc]: @johnsaigle when I add true I'm unable to access the page when using my admin account. Any idea why? I'll try debugging in the meantime.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 4]
```diff
@@ -115,11 +119,12 @@ class AjaxTimepoint extends \NDB_Page
     {
         // Verify cand ID.
         $values['identifier'] = isset($values['identifier']) ?
-            htmlentities(
+            intval(htmlentities(
```
Comment (johnsaigle) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]: You can just use `intval()` here instead of wrapping `htmlentities` because it'll only ever give you an integer.

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 5]
```diff
@@ -0,0 +1,356 @@
+import React from 'react';
+import Panel from 'Panel';
+import Loader from 'Loader';
+import PropTypes from 'prop-types';
```
Comment (zaliqarosli) [inline modules/create_timepoint/jsx/createTimepointIndex.js]: i would group react and proptypes together, then panel and loader together to make it clear what's a react package and what's a loris component

File: php/libraries/Candidate.class.inc [hunk 6]
```diff
@@ -630,43 +630,17 @@ class Candidate
...
         return intval($result);
     }
 
-    /**
-     * Returns list of consents and their respective statuses for this candidate
-     *
-     * @return array List of consents and their associated values for this candidate
-     *               The keys of the arrays are the IDs of the consents
-     */
-    public function getConsents(): array
```
Comment (zaliqarosli) [inline php/libraries/Candidate.class.inc]: why is the candidate class being touched and why is this entire function getting deleted? I'm pretty sure it's being used somewhere

**Comment by zaliqarosli** [comment]
on testing: 

Screenshot 2019-03-25 17 11 34

error messages show up on first render, but should only be enabled on submit and validation

File: php/libraries/Candidate.class.inc (hunk 6 above)
Comment (maltheism) [reply php/libraries/Candidate.class.inc]: @zaliqarosli thanks, not sure how that happened.

**Comment by maltheism** [comment]
Fixed the above, removing needs work tag.

**Comment by johnsaigle** [comment]
Integration tests still appear to need work before this can be merged.

**Comment by zaliqarosli** [comment]
error: on clicking 'Create Time Point' button 

Screenshot 2019-04-01 13 52 15

**Comment by maltheism** [comment]
Thanks @zaliqarosli, can you test again. I fixed that error I believe.

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+  }
+
+  /**
+   * Handle form submission
+   * @param {object} e - Form submission event
+   */
+  handleSubmit(e) {
+    const state = Object.assign({}, this.state);
+    state.form.options.required = {
+      subproject: true,
```
Comment (zaliqarosli) [inline modules/create_timepoint/jsx/createTimepointIndex.js]: i think this should be done in render() and not handleSubmit() because a post request is still being sent here even if fields are empty.
Screenshot 2019-04-01 14 32 20

I think a better solution to requiring fields is to set their required prop to true but turn off the error message. the html form can then throw the required error on the browser itself, i.e. ..

Screenshot 2019-04-01 14 43 14

Comment (maltheism) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: Okay i'll fix this tomorrow, have to go but note for me from @zaliqarosli is to set the values to null and not empty string.

**Comment by maltheism** [comment]
@zaliqarosli I applied the correct fix as you suggested. The form uses the default browser functionality for missing fields now.

**Comment by zaliqarosli** [comment]
@maltheism looks good! on major, only the valid subprojects show up as options. on this branch, all the subprojects show up. any chance you could make these options dynamically available like the time point options?

**Comment by maltheism** [comment]
Hi @zaliqarosli, I added the update to show valid subprojects. Please let me know if something is wrong with it.

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 8]
```diff
@@ -122,9 +122,13 @@ class CreateTimepoint extends React.Component {
...
       .then(
         (data) => {
+          console.log(data);
           // Populate the form errors.
-          if (data.errors && data.errors.length > 0) {
+          if (data.errors) {
+            // data for the form errors.
             this.setState({errors: data.errors});
+            this.setState({isLoaded: true});
+            return;
```
Comment (zaliqarosli) [inline modules/create_timepoint/jsx/createTimepointIndex.js]: why these changes?

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 9]
```diff
@@ -122,9 +122,13 @@ class CreateTimepoint extends React.Component {
     ).then((response) => response.json())
       .then(
         (data) => {
+          console.log(data);
```
Comment (zaliqarosli) [inline modules/create_timepoint/jsx/createTimepointIndex.js]: console.error is better

File: modules/create_timepoint/jsx/createTimepointIndex.js (hunk 8 above)
Comment (zaliqarosli) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: nevermind, just realized these were removed after

File: modules/create_timepoint/jsx/createTimepointIndex.js (hunk 9 above)
Comment (zaliqarosli) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: nevermind, just realized these were removed after

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 10]
```diff
@@ -122,13 +123,12 @@ class CreateTimepoint extends React.Component {
...
-          console.log(data);
           // Populate the form errors.
           if (data.errors) {
-            // data for the form errors.
-            this.setState({errors: data.errors});
-            this.setState({isLoaded: true});
-            return;
+            const state = Object.assign({}, this.state);
+            state.errors = data.errors;
+            state.form.disabled = true;
```
Comment (zaliqarosli) [inline modules/create_timepoint/jsx/createTimepointIndex.js]: could you explain what you're doing here with disabling the button? I can't trigger this case on the front-end

Comment (maltheism) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: Thanks @zaliqarosli, that's code that isn't needed anymore since your help with the unfilled form errors.

File: jsx/Form.js [hunk 11]
```diff
@@ -1496,6 +1496,7 @@ class ButtonElement extends Component {
             type={this.props.type}
             className={this.props.buttonClass}
             onClick={this.handleClick}
+            disabled={this.props.disable}
```
Comment (zaliqarosli) [inline jsx/Form.js]: while i do like this feature, i don't think it should be introduced in this PR

File: modules/create_timepoint/js/.gitignore [hunk 12]
```diff
@@ -0,0 +1 @@
+*.js
```
Comment (zaliqarosli) [inline modules/create_timepoint/js/.gitignore]: i think we're standardizing the .gitignore to be in the module directory with js/* in it

Comment (zaliqarosli) [reply modules/create_timepoint/js/.gitignore]: reference https://github.com/aces/Loris/pull/4426

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+         * Parse POST request body.
+         */
+        $requestBody = $request->getParsedBody();
+
+        /**
+         * User permission verification.
+         */
+        $user = $request->getAttribute('user');
+        if (!isset($requestBody['identifier'])) {
+            throw new LorisException(
```
Comment (johnsaigle) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]: I think this could be a `404` or `400` instead.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+                    = array_filter($labelOptions);
+            }
+        }
+        $values['visit'] = $visit_options;
+        if (!empty($visit_options)) {
+            // List of sites for the user.
+            $user = \User::singleton();
+            $DB   = \Database::singleton();
+            $user_list_of_sites = $user->getData('CenterIDs');
+            $num_sites          = count($user->getData('CenterIDs'));
```
Comment (johnsaigle) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]:
```suggestion
            $num_sites          = count($user_list_of_sites);
```

**Comment by maltheism** [comment]
Thanks @zaliqarosli and @johnsaigle. I made added the suggestions.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+    }
+
+    /**
+     * Sanitize the $postResponse data.
+     *
+     * @param array $values that have been sanitized.
+     *
+     * @return array
+     */
+    function sanitizer(array $values) : array
```
[the remark this answers is not in the document]
Comment (maltheism) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: @johnsaigle I'm a little confused how security has dibs on the word sanitizer. I typically use that word when writing a method that confirms if the receiving data is actually valid; which the function her is checking if anything is set and or make it an int.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+                    $values['visit'],
+                    $site
+                );
+                $response['status'] = 'success';
+            } else {
+                $response['status'] = 'error';
+                $response['errors'] = $errors;
+            }
+        } else {
+            $response['status'] = 'error';
```
[the remark this answers is not in the document]
Comment (maltheism) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: @johnsaigle the error message is on line 166 and is returned to the frontend in $response['errors'].

File: modules/create_timepoint/php/ajaxtimepoint.class.inc (hunk 14 above)
Comment (maltheism) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: @johnsaigle thanks will fix.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc (hunk 15 above)
[the remark this answers is not in the document]
Comment (johnsaigle) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: This function was originally doing HTML-escaping of data that would be printed to the front-end. This is done in order to prevent from JavaScript injection attacks. That was the meaning behind the `$sanitizedIdentifier` in the old code. It's not that security has dibs but that you've changed what this function is doing from previous commits.

The discussion about the HTML-escaping not being necessary anymore seems to have been lost with all the commits, but that's the context I'm coming at this from.

It's misleading to me to still have this function given that we do use the word sanitize to mean security-checks in an informal context.

Even without that though, you don't need this function. You're already using `identifer` without any check on line 78. Also both your `validate()` and `initializeSetup` do different ways of validating `psc` so you are duplicating your validation code a few times within this file. 

So given that it's potentially misleading and also not really used, I think it makes sense to delete this function.

[the remark this answers is not in the document]
Comment (maltheism) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: @johnsaigle Thanks for the in-depth writeup. Yah when I look at the code it seems weird that I didn't place the function call of validating the response data before the permissions check in any case you're right that the function seems pointless. I'll just make some tweaks to storing the identifier as an int when I do the permissions check and delete the function. Hopefully nothing breaks.

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 17]
```diff
@@ -0,0 +1,362 @@
...
+            const state = Object.assign({}, this.state);
+            state.storage.visit = data.visit;
+            this.setState(state);
+            // Handle visit selection.
+            this.handleVisitLabel();
+          }
+          // Display form to user.
+          this.setState({isLoaded: true});
+        }).catch((error) => {
+      this.populateErrors({message: 'Server error.'});
```
Comment (zaliqarosli) [inline modules/create_timepoint/jsx/createTimepointIndex.js]: i think you can remove the call to this function now

Comment (maltheism) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: @zaliqarosli I think it's good practice to have an error to display to the user by the rare chance the server responds with an error. Even if that case should only arise if the server or database goes down at the very moment a post is being sent. Thoughts?

Comment (maltheism) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: @zaliqarosli, I made a change that should resolved it.

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 18]
```diff
@@ -118,44 +118,50 @@ class CreateTimepoint extends React.Component {
...
+          this.setState(state);
+          // Handle visit selection.
+          this.handleVisitLabel();
+        }
+        // Display form to user.
+        this.setState({isLoaded: true});
+      } else {
+        response.text().then((message) => {
+          const state = Object.assign({}, this.state);
+          state.errors = [{error: message}];
```
Comment (zaliqarosli) [inline modules/create_timepoint/jsx/createTimepointIndex.js]: should this just be `state.errors = message`?

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 19]
```diff
@@ -118,44 +118,50 @@ class CreateTimepoint extends React.Component {
...
+          this.setState(state);
+        });
+      }
         }).catch((error) => {
```
Comment (zaliqarosli) [inline modules/create_timepoint/jsx/createTimepointIndex.js]: i think this line's indentation needs to be moved to the left a few

File: modules/create_timepoint/jsx/createTimepointIndex.js (hunk 18 above)
Comment (zaliqarosli) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: will the message get rendered if it is only stored into state.error here?

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 20]
```diff
@@ -118,44 +118,50 @@ class CreateTimepoint extends React.Component {
...
-      this.populateErrors({message: 'Server error.'});
+          console.log(error);
```
Comment (zaliqarosli) [inline modules/create_timepoint/jsx/createTimepointIndex.js]: console.error is better!

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 21]
```diff
@@ -118,50 +118,46 @@ class CreateTimepoint extends React.Component {
...
+            state.storage.visit = data.visit;
+            this.setState(state);
+            // Handle visit selection.
+            this.handleVisitLabel();
+          }
+          // Display form to user.
+          this.setState({isLoaded: true});
         }).catch((error) => {
-          console.log(error);
+      const state = Object.assign({}, this.state);
```
Comment (zaliqarosli) [inline modules/create_timepoint/jsx/createTimepointIndex.js]: ahh! im confused. you changed it back!

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 22]
```diff
@@ -118,50 +118,46 @@ class CreateTimepoint extends React.Component {
...
+      state.errors = [{message: error}];
```
Comment (zaliqarosli) [inline modules/create_timepoint/jsx/createTimepointIndex.js]: this is still not right i don't think. should be `state.errors = error`. or error.message even? i don't remember what the error object looks like

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 23]
```diff
@@ -118,50 +118,46 @@ class CreateTimepoint extends React.Component {
...
+      this.setState(state);
```
Comment (zaliqarosli) [inline modules/create_timepoint/jsx/createTimepointIndex.js]: the error is going into state but does it get rendered? should at least have console.error

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 24]
```diff
@@ -118,50 +118,46 @@ class CreateTimepoint extends React.Component {
...
       this.setState({isLoaded: true});
```
Comment (zaliqarosli) [inline modules/create_timepoint/jsx/createTimepointIndex.js]: you're setting state twice here back to back, could do it only once

File: modules/create_timepoint/jsx/createTimepointIndex.js (hunk 18 above)
Comment (maltheism) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: @zaliqarosli state.errors is an array for holding errors by the format of a json object and the rendering handles it as an array of json objects.

File: modules/create_timepoint/jsx/createTimepointIndex.js (hunk 19 above)
Comment (maltheism) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: @zaliqarosli The indentation is actually correct how it is. The .then().then().. makes this unpleasant on the eyes but it's actually correct.

File: modules/create_timepoint/jsx/createTimepointIndex.js (hunk 21 above)
Comment (maltheism) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: Yah I examined it for a bit and think it's better this way.

File: modules/create_timepoint/jsx/createTimepointIndex.js (hunk 22 above)
Comment (maltheism) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: @zaliqarosli see the above comment and we can chat offline if you want.

File: modules/create_timepoint/jsx/createTimepointIndex.js (hunk 23 above)
Comment (maltheism) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: The error will be displayed to the user. See line 262 and 313. It's how all the errors are shown to the user.

**Comment by kongtiaowang** [comment]
#4474
After this PR, you need to rebase your PR to pass Travis.

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 25]
```diff
@@ -0,0 +1,364 @@
...
+                {visit}
+                <ButtonElement
+                  label={'Create Time Point'}
+                  type={'submit'}
+                  name={'fire_away'}
+                />
+              </FormElement>
+            </Panel>
+          </div>
+        </div>
```
[the remark this answers is not in the document]
Comment (maltheism) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: @HenriRabalais thanks, I fixed it but seems to not be showing as outdated on Github?

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 26]
```diff
@@ -0,0 +1,364 @@
...
+    } else {
+      return (
+        <div>
+          <div>
+            <h3>New time point successfully registered.</h3>
+            <a href={'/' + this.state.url.params.candID}>
+              Click here to continue.
+            </a>
+          </div>
+        </div>
```
[the remark this answers is not in the document]
Comment (maltheism) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: @HenriRabalais thanks, I fixed it but seems to not be showing as outdated on Github?

File: modules/create_timepoint/jsx/createTimepointIndex.js (hunk 25 above)
[the remark this answers is not in the document]
Comment (HenriRabalais) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: It disappeared for me!

**Comment by maltheism** [comment]
Hi @kongtiaowang, today I had to resolve conflicting files that were merged and now some of the tests are failing in travis. I need help because after manually testing the module everything seems to be working.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 27]
```diff
@@ -0,0 +1,290 @@
...
+
+        // All subprojects from config file (error).
+        if (empty($allSubprojects)) {
+            $errors['subprojectID'] = 'No subprojects have been defined 
+            for this study. If you are an administrator, please use the 
+            Configuration module to add new subprojects.';
+        }
+
+        // List of valid subprojects for a given project
+        if ($config->getSetting('useProjects') === 'true') {
```
Comment (kongtiaowang) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]: we don't use 'useProjects' anymore.

Comment (kongtiaowang) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: https://github.com/aces/Loris/pull/4496/

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 28]
```diff
@@ -0,0 +1,284 @@
...
+                    $labelOptions[$item['@']['value']] = $item['#'];
+                }
+                $visit_options[$visitLabel['@']['subprojectID']]
+                    = array_filter($labelOptions);
+            }
+        }
+        $values['visit'] = $visit_options;
+        if (!empty($visit_options)) {
+            // List of sites for the user.
+            $user = \User::singleton();
```
Comment (xlecours) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]: This whole block could be replaced by:
```
$values['psc'] = array_intersect(\Utility::getSiteList(), \User::singleton()->getSiteNames());
```

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 29]
```diff
@@ -0,0 +1,284 @@
...
+
+    /**
+     * Initialize setup, the extra values for the
+     * create timepoint form. (psc & errors)
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     */
+    function initializeSetup(array $values) : array
```
Comment (xlecours) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]: suggestion :
`private function _handleGET(ServerRequestInterface $request): ResponseInteface`

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 30]
```diff
@@ -0,0 +1,284 @@
...
+     * Processes the values and saves to database
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     */
+    function processRequest(array $values) : array
+    {
+        if (isset($values['command']) && $values['command'] == 'initialize') {
+            $response = $this->initializeSetup($values);
```
Comment (xlecours) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]: This should not be a POST. If the goal is to get values from the server then GET should be used.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 31]
```diff
@@ -0,0 +1,284 @@
...
+
+        // List of valid subprojects for a given project
+        $subprojList = $candidate->getValidSubprojects();
+        if (empty($subprojList)) {
+            $this->form->errors['subprojectID'] = "No subprojects have been 
+            defined for the project this candidate is affiliated with. 
+            If you are an administrator, please use the Configuration module to 
+            add new subprojects and associate them with projects.";
+        }
+        $tmp = array();
```
Comment (xlecours) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]: `$values['subproject'] = array_intersect_key(\Utility::getSubprojectList(), $candidate->getValidSubprojects());`

**Comment by maltheism** [comment]
> Here are some change requests.
> General comment: I think that `GET ...create_timepoint/AjaxTimepoint` should be used to access initial values for the drop down.

Hey @xlecours The reason I'm using a POST is because the method contains data that's being sent (ex. 'initialize' or 'create') for retrieval and GET contains no data with the request.

**Comment by xlecours** [comment]
> > Here are some change requests.
> > General comment: I think that `GET ...create_timepoint/AjaxTimepoint` should be used to access initial values for the drop down.
> 
> Hey @xlecours The reason I'm using a POST is because the method contains data that's being sent (ex. 'initialize' or 'create') for retrieval and GET contains no data with the request.

@maltheism But you could assume that a GET request is for 'initialize' and POST for 'create'.

**Comment by maltheism** [comment]
> > > Here are some change requests.
> > > General comment: I think that `GET ...create_timepoint/AjaxTimepoint` should be used to access initial values for the drop down.
> > 
> > 
> > Hey @xlecours The reason I'm using a POST is because the method contains data that's being sent (ex. 'initialize' or 'create') for retrieval and GET contains no data with the request.
> 
> @maltheism But you could assume that a GET request is for 'initialize' and POST for 'create'.

@xlecours Okay, I'll refactor it to do that once I fix the PR. Since it seems to be broken as of today.

**Comment by maltheism** [comment]
Hey @xlecours, I added the changes you suggested. 🙂

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 32]
```diff
@@ -0,0 +1,285 @@
...
+            case 'Forbidden':
+                return new \LORIS\Http\Response\Forbidden(
+                    'Forbidden'
+                );
+            default:
+                break;
+            }
+        }
+
+        if (isset($values['command']) && $values['command'] == 'create') {
```
Comment (xlecours) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]: the command parameter is not necessary. A POST request to this endpoint means that a timepoint should be created with the given data.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 33]
```diff
@@ -0,0 +1,285 @@
...
+
+    /**
+     * Permission verification.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     * @param array                  $values  The form values.
+     *
+     * @return array
+     */
+    private function verifyPermissions(
```
Comment (xlecours) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]: By convention, there should be `private function _hasAccess(\User $user): bool` that only check for permission.
Then, in the handlePOST, check for mandatory fields and return a BadRequest when one is missing.
Then, instantiate the candidate and compare its RegistrationCenterId to the user's sites. 

In it's current state, the `verifyPermissions` function is doing all this and have a side effect of modifying the values

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 34]
```diff
@@ -0,0 +1,285 @@
...
+        }
+
+        if (isset($values['command']) && $values['command'] == 'create') {
+            $errors = $this->validate($values);
+            // convert site entered to a \Site object
+            $site = null;
+            $user = \User::singleton();
+            $user_list_of_sites = $user->getCenterIDs();
+            $num_sites          = count($user_list_of_sites);
+            if ($num_sites == 1) {
```
Comment (xlecours) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]: Could you make the React app always send the Site even when the user as only one? Maybe just don't display the fields when it's the case.
We would then be able to avoid this "auto-select" feature in the backend.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 35]
```diff
@@ -0,0 +1,285 @@
...
+                return new \LORIS\Http\Response\Forbidden(
+                    'Forbidden'
+                );
+            default:
+                break;
+            }
+        }
+
+        if (isset($values['command']) && $values['command'] == 'create') {
+            $errors = $this->validate($values);
```
Comment (xlecours) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]: if there is an error, `return new \LORIS\Http\Response\BadRequest` right away. 
Doing this prevent from having to do special `if` statement like at line#193

File: modules/create_timepoint/php/ajaxtimepoint.class.inc [hunk 36]
```diff
@@ -0,0 +1,285 @@
...
+                );
+                $response['status'] = 'success';
+            } else {
+                $response['status'] = 'error';
+                $response['errors'] = $errors;
+            }
+        } else {
+            $response['status'] = 'error';
+        }
+        return new \LORIS\Http\Response\JsonResponse(
```
Comment (xlecours) [inline modules/create_timepoint/php/ajaxtimepoint.class.inc]: (Comment/ not a formal change request) :: It would be nice if this was a 201 with the location header equal to baseurl + /candid so that the React app just need to send `window.lcoation($url)` instead of the success page with the link to click.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/201

File: modules/create_timepoint/jsx/createTimepointIndex.js [hunk 37]
```diff
@@ -0,0 +1,359 @@
...
+  /**
+   * Retrieve data from the provided URL and save it in state.
+   */
+  fetchInitializerData() {
+    const state = Object.assign({}, this.state);
+    let url = new URL(this.props.dataURL);
+    const params = {
+      candID: state.url.params.candID,
+      identifier: state.url.params.identifier,
+      subprojectID: state.form.subproject,
```
[the remark this answers is not in the document]
Comment (maltheism) [reply modules/create_timepoint/jsx/createTimepointIndex.js]: @xlecours I think you're right. I cannot remember why I included it for the request.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc (hunk 35 above)
Comment (maltheism) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: @xlecours I think the error message displayed on the frontend might be more helpful than the BadRequest?

File: modules/create_timepoint/php/ajaxtimepoint.class.inc (hunk 33 above)
Comment (maltheism) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: @xlecours this function is using the code before I modified the module for react. The $candidate is made from the $values['identifier] and that's used in the hasPermission('data_entry') check. Do you want me to move the validating the identifier sent from the fronted out of this function?

File: modules/create_timepoint/php/ajaxtimepoint.class.inc (hunk 36 above)
Comment (maltheism) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: @xlecours you want me to skip the "New time point successfully registered" and just redirect to the /candID link automatically? I can do that by just checking with the candID gets updated.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc (hunk 33 above)
Comment (xlecours) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: @maltheism yes, I would like that.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc (hunk 34 above)
Comment (xlecours) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: @maltheism I the POST request body, there should be a site.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc (hunk 35 above)
Comment (xlecours) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: you can add any message to a BadResquest response.

File: modules/create_timepoint/php/ajaxtimepoint.class.inc (hunk 34 above)
Comment (maltheism) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: @xlecours you're right. I'm still confused on what you want me to do because won't I still need to use $site = \Site::singleton($user_list_of_sites[0]); for retrieving the site object that's used for creating the timepoint?

Comment (xlecours) [reply modules/create_timepoint/php/ajaxtimepoint.class.inc]: `$site = \Site::singleton(intval($values['psc']));` would be enough. Because `$values` will always have a `psc` otherwise a BadRequest would have been sent earlier.

the only scenario where this would fail is if the user is only part of a study site. The array intersect at line#136. in the handleGET, would not send a psc to the front end because Utility::getSiteList would remove the DCC sites from the list. Which would create a weird behavior with `$site = \Site::singleton($user_list_of_sites[0]);` (I would register the session as part of the study_site)

**Comment by xlecours** [comment]
Regarding Travis,
There were 2 errors:
```
1) CreateTimepointTestIntegrationTest::testCreateTimepoint
UnknownServerException: input is null
Build info: version: '2.53.1', revision: 'a36b8b1', time: '2016-06-30 17:37:03'
System info: host: '274ed7ad209e', ip: '172.18.0.2', os.name: 'Linux', os.arch: 'amd64', os.version: '4.15.0-1028-gcp', java.version: '1.8.0_91'
Driver info: driver.version: unknown
/app/vendor/facebook/webdriver/lib/WebDriverExceptions.php:79
/app/vendor/facebook/webdriver/lib/remote/HttpCommandExecutor.php:268
/app/vendor/facebook/webdriver/lib/remote/RemoteWebDriver.php:491
/app/vendor/facebook/webdriver/lib/remote/RemoteWebDriver.php:270
/app/modules/create_timepoint/test/create_timepointTest.php:104
/app/modules/create_timepoint/test/create_timepointTest.php:79
```
It is a Javascript error from line 104 of create_timepointTest.php. The querySelector returns nothing for `subproject, select`. Try changing the value of `static $subProject` to `'select[name="subproject"]'`. It can be tested in the browser console `document.querySelector('select[name="subproject"]')`

```
2) CreateTimepointTestIntegrationTest::testCreateTimepointErrorEmptySubproject
NoSuchElementException: Unable to locate element: {"method":"name","selector":"fire_away"}
For documentation on this error, please visit: http://seleniumhq.org/exceptions/no_such_element.html
Build info: version: '2.53.1', revision: 'a36b8b1', time: '2016-06-30 17:37:03'
System info: host: '274ed7ad209e', ip: '172.18.0.2', os.name: 'Linux', os.arch: 'amd64', os.version: '4.15.0-1028-gcp', java.version: '1.8.0_91'
Driver info: driver.version: unknown
/app/vendor/facebook/webdriver/lib/WebDriverExceptions.php:67
/app/vendor/facebook/webdriver/lib/remote/HttpCommandExecutor.php:268
/app/vendor/facebook/webdriver/lib/remote/RemoteWebDriver.php:491
/app/vendor/facebook/webdriver/lib/remote/RemoteWebDriver.php:133
/app/modules/create_timepoint/test/create_timepointTest.php:133
```
Since it's now a React app, the button is rendered after the page load. The current WebdriverBy call at line#133 should be change to execute script.
```
document.querySelector('button[name="fire_away"]').click();
```

**Comment by maltheism** [comment]
Closing because of failed rebase. See #5238
